### PR TITLE
Feature/gh 13536 saved cart e2e improvements

### LINKED ISSUE #13536: E2E Tests for b2b saved cart
Looking at the feature of b2b saved cart feature, I think we could add these e2e tests to increase the test coverage.

Test scenarios proposed:

-> Updating quantity in a saved cart
     - Have a saved cart with a few items, for eg, 3 different items
     - Go to the saved cart
     - Update the quantity of one of the items
     - Remove one item from the saved cart 
     - Make the cart active
     - Proceed to checkout

-> Clone a saved cart twice
    - Have a cart
    - Save the cart, with a name, for eg nameA.
    - Make the saved cart active and clone it (so now the cloned cart is named Copy of nameA)
    - Go to cart, and save the cart again, with the same name, nameA
    - Make the saved cart active and clone it (so the cloned cart is also named Copy of nameA)
    - Verify that the ID is different and unique for the cloned carts with same name. 

-> Make saved cart active through saved cart details page
    - Have a couple of saved carts.
    - Select one saved cart to go to its saved cart details.
    - Make the saved cart active from there

### PR DESCRIPTION
**Pull request by dydome**
close #13536

### CODE REVIEW

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts [hunk 1]
```diff
@@ -100,6 +100,33 @@ context('B2B - Saved Cart', () => {
...
+        it('should create cart, save it and restore with a clone with default form values twice and then compare row in saved cart listing page', () => {
+          savedCart.saveActiveCart();
+
+          cy.get('cx-saved-cart-list .cx-saved-cart-list-table tr')
+            .first()
+            .find('.cx-saved-cart-list-cart-id > .cx-saved-cart-list-value')
+            .then((cartId) => {
+              const cartCode = cartId.text();
+
+              savedCart.manuallyRestoryCartFromListingPage(cartCode);
```
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts]: you should assert that you have a cloned saved cart afterwords in the listing page

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts [hunk 2]
```diff
@@ -100,6 +100,33 @@ context('B2B - Saved Cart', () => {
...
+              savedCart.visitCartPage();
+
+              cy.window()
+                .then((win) =>
+                  JSON.parse(
+                    win.localStorage.getItem('spartacus⚿powertools-spa⚿cart')
+                  )
+                )
+                .then(({ activeCartCode }) => {
+                  savedCart.saveActiveCart(true, false);
```
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts]: what's the point of passing true for manual and false for checking the alert. When we save it, it shows so why not check it?

Also, what's the point of these params to check? We should always assert when we save a cart that it shows the global message and was saved

--------
Since the saveActiveCart function is asserting from another place, maybe you can update the function to be more 'dynamic' when doing the assertions. I'll leave it up to Kate to decide, but either way would be fine as you have already created other functions to 'bypass' this.

Comment (bgambocjaviniar) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts]: Basically, for this case, there should have 2 saved carts in the listing page now. It would be great to check if it was saved and that there are 2 saved carts now

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-saved-cart.ts [hunk 3]
```diff
@@ -302,6 +404,55 @@ export function verifyCartDetails(cart: any) {
...
+  cy.get(
+    'cx-saved-cart-form-dialog .cx-saved-cart-form-footer .btn-primary'
+  ).click();
+
+  cy.wait(`@${restoreSavedCartAlias}`)
+    .its('response.statusCode')
+    .should('eq', 200);
+}
+
+export function manuallyRestoryCartFromListingPage(cartCode: string) {
```
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-saved-cart.ts]: manuallyRestoryCartFromListingPage --> manuallyRestoreCartFromListingPage

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts [hunk 4]
```diff
@@ -178,6 +205,50 @@ context('B2B - Saved Cart', () => {
           sampleData.savedActiveCartForm[0]
         );
       });
+
+      it('should make saved cart active from details page', () => {
+        savedCart.waitForSavedCartListingPageData([
```
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts]:
```suggestion
        savedCart.waitForSavedCartListingPageData(sampleData.products)
```

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts [hunk 5]
```diff
@@ -178,6 +205,50 @@ context('B2B - Saved Cart', () => {
...
+        cy.get('cx-saved-cart-list .cx-saved-cart-list-table tr')
+          .first()
+          .find('.cx-saved-cart-list-cart-id > .cx-saved-cart-list-value')
+          .then((cartId) => {
+            const cartCode = cartId.text();
+
+            savedCart.openFirstElementOnTheList(cartCode);
+            savedCart.manuallyRestoryCartFromDetailsPage(cartCode);
+            savedCart.verifyMiniCartQuantity(1);
+            savedCart.visitSavedCartListingPage();
```
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts]: There's no point to have this as when you restore the saved cart from the details page, it automatically redirects you to the listing page

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-saved-cart.ts [hunk 6]
```diff
@@ -302,6 +404,55 @@ export function verifyCartDetails(cart: any) {
     });
 }
 
+export function manuallyRestoryCartFromDetailsPage(cartCode: string) {
```
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-saved-cart.ts]: manuallyRestoryCartFromDetailsPage --> manuallyRestoreCartFromDetailsPage

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts [hunk 7]
```diff
@@ -178,6 +205,50 @@ context('B2B - Saved Cart', () => {
...
+          .then((cartId) => {
+            const cartCode = cartId.text();
+
+            savedCart.openFirstElementOnTheList(cartCode);
+            savedCart.verifySavedCartDetailsItemsQuantity(2);
+
+            savedCart.removeFirstProductFromSavedCart(cartCode);
+            savedCart.verifySavedCartDetailsItemsQuantity(1);
+
+            savedCart.modifyFirstProductQuantityFromSaveCart(2, cartCode);
```
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts:244]: I don't think you have the last 2 of what was mentioned below. I would discuss with Kate if we should really should checkout

```
-> Updating quantity in a saved cart
- Have a saved cart with a few items, for eg, 3 different items
- Go to the saved cart
- Update the quantity of one of the items
- Remove one item from the saved cart
- Make the cart active
- Proceed to checkout
```

**Review by bgambocjaviniar — CHANGES_REQUESTED** [review]
Did a general pass as requested as you mentioned Kate will review

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts (hunk 7 above)
Comment (dydome) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts:244]: Asked Kate

Comment (dydome) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts:244]: I have added 'make cart active' but proceed to checkout is nowhere in the saved cart e2e file so let's leave it.

Comment (KateChuen) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts:244]: Yep, I agreed to have checkout on a separate saved cart test (written below), as long as we complete an order on a saved cart.

File: projects/storefrontapp/src/environments/environment.ts [hunk 8]
```diff
@@ -18,7 +18,8 @@ export const environment: Environment = {
   occBaseUrl: buildProcess.env.CX_BASE_URL,
   occApiPrefix: '/occ/v2/',
   cds: buildProcess.env.CX_CDS ?? false,
-  b2b: buildProcess.env.CX_B2B ?? false,
+  b2b: true,
```
Comment (KateChuen) [inline projects/storefrontapp/src/environments/environment.ts]: don't forget to revert the changes in this file

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts [hunk 9]
```diff
@@ -178,6 +208,55 @@ context('B2B - Saved Cart', () => {
...
+      it('should perform full checkout from saved cart details', () => {
+        savedCart.updateSavedCartAndRestore(
+          sampleData.products[0],
+          sampleData.savedActiveCartForm[0]
+        );
+
+        // Checkout flow
+        savedCart.proceedToCheckout();
+        savedCart.executeCheckout();
+      });
```
Comment (KateChuen) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/saved-cart/b2b-saved-cart.e2e-spec.ts:259]: can we add an assertion to make sure the item on the order confirmation page is the same as the one in the saved cart?

**Review by KateChuen — DISMISSED** [review]

**Comment by bgambocjaviniar** [comment]
I would revert all the things after my last comment as I believe those random 'improvements' broke the e2es as I recall it was always green before. But now with the inclusion of 'hard-coded values, we have broken e2es

Starting with `Next update for e2e saved carts`, we should revert

After the revert of all those things, I would include the changes in https://github.com/SAP/spartacus/pull/14882. I don't know why they're separate. 

Screen Shot 2022-02-06 at 10 01 25 PM
